src/lib.rs

```diff
@@ -1,203 +1,960 @@
+        T::read(&t, _io, root, parent)?;
+        Ok(t)
+    }
+
+    fn downcast<T, U>(opt_rc: Option<SharedType<U>>, t: OptRc<T>, panic: bool) -> SharedType<U>
+    where
+        T: KStruct + Default + Any,
+        U: 'static,
+    {
+        if let Some(rc) = opt_rc {
+            rc
+        } else {
+            let t_any = &t.get() as &dyn Any;
+            //println!("`{}` is a '{}' type", type_name_of_val(&t), type_name::<Rc<U>>());
+            match t_any.downcast_ref::<Rc<U>>() {
+                Some(as_result) => SharedType::<U>::new(Rc::clone(as_result)),
+                None => {
+                    if (panic) {
+                        #[cfg(feature = "type_name_of_val")]
+                        panic!(
+                            "`{}` is not a '{}' type",
+                            std::any::type_name_of_val(&t),
+                            type_name::<Rc<U>>()
+                        );
+                        #[cfg(not(feature = "type_name_of_val"))]
+                        panic!("`{:p}` is not a '{}' type", &t, type_name::<Rc<U>>());
+                    }
+                    SharedType::<U>::empty()
+                }
             }
-        );
+        }
     }
+}
 
-    test_read_integer!(read_u1, 1);
-    test_read_integer!(read_s1, 1);
+/// Dummy struct used to indicate an absence of value; needed for
+/// root structs to satisfy the associated type bounds in the
+/// `KStruct` trait.
+#[derive(Debug, Default, Copy, Clone, PartialEq)]
+pub struct KStructUnit;
 
-    test_read_integer!(read_s2le, 513);
-    test_read_integer!(read_s2be, 258);
-    test_read_integer!(read_u2le, 513);
-    test_read_integer!(read_u2be, 258);
+impl KStruct for KStructUnit {
+    type Root = KStructUnit;
+    type Parent = KStructUnit;
 
-    test_read_integer!(read_s4le, 67305985);
-    test_read_integer!(read_s4be, 16909060);
-    test_read_integer!(read_u4le, 67305985);
-    test_read_integer!(read_u4be, 16909060);
+    fn read<S: KStream>(
+        _self_rc: &OptRc<Self>,
+        _io: &S,
+        _root: SharedType<Self::Root>,
+        _parent: SharedType<Self::Parent>,
+    ) -> KResult<()> {
+        Ok(())
+    }
+}
 
-    test_read_integer!(read_s8le, 578437695752307201);
-    test_read_integer!(read_s8be, 72623859790382856);
-    test_read_integer!(read_u8le, 578437695752307201);
-    test_read_integer!(read_u8be, 72623859790382856);
+use std::{fs, path::Path};
 
-    #[test]
-    fn read_f4le() {
-        let mut buf = Cursor::new(vec![0, 0, 128, 62]);
-        assert_eq!(buf.read_f4le().unwrap(), 0.25);
+impl From<std::io::Error> for KError {
+    fn from(err: std::io::Error) -> Self {
+        Self::IoError {
+            desc: err.to_string(),
+        }
     }
+}
 
-    #[test]
-    fn read_f4be() {
-        let mut buf = Cursor::new(vec![62, 128, 0, 0]);
-        assert_eq!(buf.read_f4be().unwrap(), 0.25);
+pub trait KStream {
+    fn clone(&self) -> BytesReader;
```

Comment:
I don't really understand the purpose of this trait. In theory, it makes sense to allow alternative implementations of `KStream`, but the generated code can only work with `BytesReader`. In fact, the generated code only uses this trait to call the `clone()` method, which is required here to return `BytesReader`.

Unfortunately there are such illogical things in your code everywhere... (and this one is still relatively benign compared to others). I don't know if we can ever fix them all.